File: src/adapters/registry.ts
import { HeadroomAdapter } from './headroom.js';
import { RtkAdapter } from './rtk.js';
import type { ProxyAdapter, ProxyName } from './types.js';

/**
 * Small named registry mapping the two known, locked v0.1 proxies to an
 * adapter instance -- NOT a generic plugin-loader. A config-driven or
 * dynamic-loading N-proxy plugin system would be premature abstraction at
 * this scope (2 known targets, not 30). Adding a new proxy means adding one
 * file and one line here, per CONTRIBUTING.md.
 *
 * `getAdapter('headroom')` below still constructs a real HeadroomAdapter if
 * called directly, but `runVerify()`'s v0.1 dispatch loop never actually
 * calls it -- see src/verify.ts / src/adapters/headroom.ts.
 */
const FACTORIES: Record<ProxyName, () => ProxyAdapter> = {
  rtk: () => new RtkAdapter(),
  headroom: () => new HeadroomAdapter(),
};

export const SUPPORTED_PROXIES: ProxyName[] = ['rtk', 'headroom'];

export function isSupportedProxy(name: string): name is ProxyName {
  return (SUPPORTED_PROXIES as string[]).includes(name);
}

export function getAdapter(name: ProxyName): ProxyAdapter {
  const factory = FACTORIES[name];
  if (!factory) {
    throw new Error(
      `Unknown proxy "${name}". Supported proxies: ${SUPPORTED_PROXIES.join(', ')}.`,
    );
  }
  return factory();
}

File: python/src/tokentrust/adapters/registry.py
"""Ported from src/adapters/registry.ts."""

from __future__ import annotations

from typing import Dict, List

from .headroom import HeadroomAdapter
from .rtk import RtkAdapter
from .types import ProxyAdapter, ProxyName

SUPPORTED_PROXIES: List[ProxyName] = ["rtk", "headroom"]

_FACTORIES: Dict[str, "type"] = {
    "rtk": RtkAdapter,
    "headroom": HeadroomAdapter,
}


def is_supported_proxy(name: str) -> bool:
    return name in SUPPORTED_PROXIES


def get_adapter(name: ProxyName) -> ProxyAdapter:
    factory = _FACTORIES.get(name)
    if factory is None:
        raise ValueError(f'Unknown proxy "{name}". Supported proxies: {", ".join(SUPPORTED_PROXIES)}.')
    return factory()
